oshi-core/src/main/java/oshi/software/common/AbstractOSThread.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * MIT License
+ *
+ * Copyright (c) 2010 - 2020 The OSHI Project Contributors: https://github.com/oshi/oshi/graphs/contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package oshi.software.common;
+
+import oshi.software.os.OSProcess;
+import oshi.software.os.OSThread;
+
+import java.util.Collections;
+import java.util.List;
+
+public abstract class AbstractOSThread implements OSThread {
+
+    private OSProcess parentProcess;
+
+    public AbstractOSThread(OSProcess process) {
+        this.parentProcess = process;
+    }
+
+    @Override
+    public OSProcess getParentProcess() {
+        return this.parentProcess;
+    }
+
+    @Override
+    public int getProcessID() {
+        return getParentProcess().getProcessID();
+    }
+
+    @Override
+    public int getParentProcessID() {
+        return getParentProcess().getParentProcessID();
```

Comment:
Continuation of above discussion. It seems odd to worry about a parent of a process when we are in a thread of that process. It seems to me the "parent" of the thread should be the process it is a thread of, so this should return the "owning" process ID, and the getProcessID should return the thread ID (or 0 if such a thing is undefined).  Again, not sure what the right answer is but we should make a design decision here.

Reply:
I think i did a mistake naming the process of the thread to be `parentProcess`. It should be just `process`, and hence the method `getParentProcess` should be renamed to `getProcess` to avoid confusion. But you are right, this sounds more logical to return processId in `getParentProcessID()` and 0 in `getProcessID` . What do you think?

Reply:
If you are inheriting you can’t really rename the method but you can redefine what it means.  

I think the “parent” of a thread is logically it’s process and the “processid” is the thread I’d, if it exists.

Reply:
True, so i will add `getProcessID()` to return 0 as we already have a method `getThreadId` to return threadId. Also `getParentProcessID()` will now return processId, which i think makes sense after reading your comment.

Reply:
Or just have getProcessId return the thread id and don’t have a separate thread Id method.

Reply:
ok but i still think that it should have a method to return threadId

Reply:
Are there any cases where the thread id is not a “lightweight” process id?  The problem with a separate “thread” (this object) and parent (owning process) is that the “process id” becomes ambiguous.  

If there are always only 2 numbers to ever return and we have sensibly named methods to return those 2 numbers, I don’t see added value in adding a new thread id method and also returning 0 for the process id.  I’m not strongly opposed to a new thread Id method that is just an alias to the process id, though.

Reply:
OK, after reading up on this for about an hour, I think I have this sorted.  Here's some background:
 - The Windows [THREADENTRY32](https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/ns-tlhelp32-threadentry32) includes both a thread ID and an "Owning Process ID".  So here it's not clear whether we want the pid or ppid as owner, but we are free to choose.
 - Linux thread info retains a reference to the "parent PID" so there's no reason not to return the same thing here, so let's keep parent process ID pointing to the parent process's parent, and not the "owning" process.  This means we need process ID as the owner.  
 - FreeBSD and Solaris have, in addition to a thread ID, a thread name. This would be useful info. Let's add a threadName field (default empty string).
 - Solaris thread ID is not unique (basically it's 1, 2, 3, etc. for each process). So keeping the process ID as part of a thread's identity (separate from the thread ID) is a must.
 - Mac threads don't have their own numbers easily obtainable so we can ignore them.  I'll keep digging around whatever macOS source I can find to see if there's anything useful.

So, basically:
 - keep the "new" thread ID field
 - also add a new threadName field to populate with info on FreeBSD/Solaris
 - have getProcessID return the "owning" process
 - have getParentProcessID return the owning process's parent

Reply:
Thanks for the explanation and very good points indeed. Few questions -
1. For name, there is a method `getName` which is returning `getParentProcess().getName()` for Linux. Do you want to change this to return "" for linux and return certain value for FreeBSD/Solaris?

Reply:
Also how did you test this for multiple OS. I work on ubuntu and hence could test Linux easily. Do you create VM and test it there?

Reply:
Yeah, we can probably reuse the getName field where appropriate. Leave blank except for where it's filled in.   

And for multiple OS's, yes, I mostly work on my Mac, with VMs for Win7, Ubuntu, Solaris, and FreeBSD. I also have a physical Win10 machine for work that I test windows code on.  I recently got a Synology Diskstation that supposedly has VM support and am trying to figure out how to set that up as well, but haven't gotten very far.

And I just committed the Windows mappings for `Thread32First` and `Thread32Next` to your fork/branch.

Reply:
FYI, dug in to macOS source code. `ps -M` is the closest it gets to listing threads, and all it really does is iterate from `j=0` to `j<numthreads` using `j` as the thread number and copying all the process info.  Thread 0 gets all the usage stats and threads 1-(numthreads-1) get 0's.  So it's basically useless, but I don't feel bad doing the same thing. :)
```
$ ps -aM           
USER           PID   TT   %CPU STAT PRI     STIME     UTIME COMMAND
root          1296 s004    0.0 S    31T   0:00.02   0:00.02 login -pfl danielwiddis /bin/bash -c exec -la zsh /bin/zsh
              1296         0.0 S    31T   0:00.00   0:00.00 
danielwiddis  1301 s004    0.0 S    31T   0:00.64   0:00.57 -zsh
root         60632 s004    0.0 R    31T   0:00.00   0:00.00 ps -aM
```

Reply:
Thanks for the commit for windows. I don't both windows and macOS unfortunately, but i have created VM for both solaris and FreeBSD and can work on it.

Reply:
No problem.   Look for further Windows commits :)